### PR TITLE
Cleanup run through the media backend

### PR DESCRIPTION
**Pull request by bzeller**
First cleanup run to make the Media Backend a bit leaner by getting rid of redundant layers that are not exposed to the outside
world anyway. This should also make it easier to introduce async functions because the API is more clean.

### CODE REVIEW

File: zypp/OnMediaLocation.cc [hunk 1]
```diff
@@ -145,6 +147,11 @@ namespace zypp
   OnMediaLocation & OnMediaLocation::setHeaderChecksum( CheckSum val_r )
   { _pimpl->_headerChecksum = val_r; return *this; }
 
+  const zypp::filesystem::Pathname &OnMediaLocation::deltafile() const
```
Comment (mlandres) [inline zypp/OnMediaLocation.cc:150]:
```suggestion
  const Pathname &OnMediaLocation::deltafile() const
```

File: zypp/OnMediaLocation.cc [hunk 2]
```diff
@@ -145,6 +147,11 @@ namespace zypp
...
+  { return _pimpl->_deltafile; }
+
+  OnMediaLocation &OnMediaLocation::setDeltafile(const zypp::filesystem::Pathname &path)
```
Comment (mlandres) [inline zypp/OnMediaLocation.cc:153]:
```suggestion
  OnMediaLocation &OnMediaLocation::setDeltafile(const Pathname &path)
```

File: zypp/OnMediaLocation.h [hunk 3]
```diff
@@ -116,6 +116,12 @@ namespace zypp
     /** Set the \ref headerChecksum. */
     OnMediaLocation & setHeaderChecksum( CheckSum val_r );
 
+    /** The existing deltafile that can be used to reduce download size ( zchunk or metalink ) */
+    const Pathname & deltafile() const;
+    /** Set the \ref deltafile. */
+    OnMediaLocation & setDeltafile( const Pathname &path );
```
Comment (mlandres) [inline zypp/OnMediaLocation.h:122]:
```suggestion
    OnMediaLocation & setDeltafile( Pathname path );
```
And move assign in the implementation.

File: zypp/media/MediaHandler.cc [hunk 4]
```diff
@@ -1222,46 +1221,28 @@ std::ostream & operator<<( std::ostream & str, const MediaHandler & obj )
...
-    getFile( req.first, req.second );
+  if (info.isExist())
+    ZYPP_THROW(MediaNotAFileException(url(), localPath(file.filename())));
+  else
+    ZYPP_THROW(MediaFileNotFoundException(url(), file.filename()));
 }
 
-
-void MediaHandler::getFileCopy (const Pathname & srcFilename, const Pathname & targetFilename , const ByteCount &expectedFileSize_r) const
+void MediaHandler::getFileCopy( const OnMediaLocation &file, const zypp::filesystem::Pathname &targetFilename ) const
```
Comment (mlandres) [inline zypp/media/MediaHandler.cc:1237]:
```suggestion
void MediaHandler::getFileCopy( const OnMediaLocation &file, const Pathname &targetFilename ) const
```

File: zypp/media/MediaHandlerFactory.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+#include <zypp/zyppng/media/MediaNetwork>
+
+namespace zypp::media {
+
+  MediaHandlerFactory::MediaHandlerFactory()
+  {
+
+  }
+
+  std::unique_ptr<MediaHandler> MediaHandlerFactory::createHandler( const Url &o_url, const zypp::filesystem::Pathname &preferred_attach_point )
```
Comment (mlandres) [inline zypp/media/MediaHandlerFactory.cc:28]:
```suggestion
  std::unique_ptr<MediaHandler> MediaHandlerFactory::createHandler( const Url &o_url, const Pathname &preferred_attach_point )
```

File: zypp/media/MediaManager.cc [hunk 6]
```diff
@@ -36,37 +37,91 @@ namespace zypp
...
+          , verifier( std::move(m.verifier) )
+          , _handler ( std::move(m._handler) )
         {}
 
-        ManagedMedia(const ManagedMedia &m)
-          : desired (m.desired)
-          , handler (m.handler)
-          , verifier(m.verifier)
-        {}
+        static ManagedMedia makeManagedMedia ( const Url &o_url, const zypp::filesystem::Pathname &preferred_attach_point, const MediaVerifierRef &v )
```
Comment (mlandres) [inline zypp/media/MediaManager.cc:61]:
```suggestion
        static ManagedMedia makeManagedMedia ( const Url &o_url, const Pathname &preferred_attach_point, const MediaVerifierRef &v )
```
+ all the cases I missed to see ;)

File: zypp/zyppng/media/medianetwork.h [hunk 7]
```diff
@@ -45,8 +45,7 @@ class MediaHandlerNetwork : public zypp::media::MediaHandler
 protected:
   void attachTo(bool next) override;
   void releaseFrom(const std::string &ejectDev) override;
-  void getFile(const zypp::filesystem::Pathname &filename, const zypp::ByteCount &expectedFileSize_r) const override;
-  void getFiles(const std::vector<std::pair<zypp::filesystem::Pathname, zypp::ByteCount> > &files) const override;
+  void getFile ( const zypp::OnMediaLocation &file ) const override;
   void getDir(const zypp::filesystem::Pathname &dirname, bool recurse_r) const override;
```
Comment (mlandres) [inline zypp/zyppng/media/medianetwork.h:49]: `find zypp/ -type f | xargs -d '\n' sed -i 's/zypp::filesystem::Pathname/Pathname/g'` maybe...

**Review by mlandres — APPROVED** [review]
Minor nagging, but good to be merged.

File: zypp/OnMediaLocation.cc (hunk 1 above)
Comment (bzeller) [reply zypp/OnMediaLocation.cc:150]: Ok

File: zypp/OnMediaLocation.cc (hunk 2 above)
Comment (bzeller) [reply zypp/OnMediaLocation.cc:153]: Ok

File: zypp/OnMediaLocation.h (hunk 3 above)
Comment (bzeller) [reply zypp/OnMediaLocation.h:122]: Ok

File: zypp/media/MediaHandler.cc (hunk 4 above)
Comment (bzeller) [reply zypp/media/MediaHandler.cc:1237]: Ok

File: zypp/media/MediaHandlerFactory.cc (hunk 5 above)
Comment (bzeller) [reply zypp/media/MediaHandlerFactory.cc:28]: Ok

File: zypp/media/MediaManager.cc (hunk 6 above)
Comment (bzeller) [reply zypp/media/MediaManager.cc:61]: Ok

File: zypp/zyppng/media/medianetwork.h (hunk 7 above)
Comment (bzeller) [reply zypp/zyppng/media/medianetwork.h:49]: That does not work in zyppng, different namespace. We could  pull Pathname into the zyppng namespace , but that would be a seperate commit and would touch a lot of unrelated code here

**Pull request merged**